omero_figure/scripts/omero/figure_scripts/Figure_To_Pdf.py

```diff
@@ -1109,6 +1109,140 @@ def draw_scalebar(self, panel, region_width, page):
+        x = int(x * scale)
+        y = int(y * scale)
+        width = int(width * scale)
+        height = int(height * scale)
+        size_x = int(size_x * scale)
+        size_y = int(size_y * scale)
+
+        canvas = None
+        # Coordinates below are all final jpeg coordinates & sizes
+        if x < 0 or y < 0 or (x + width) > size_x or (y + height) > size_y:
+            # If we're outside the bounds of the image...
+            # Need to render reduced region and paste on to full size image
+            canvas = Image.new("RGBA", (width, height), (221, 221, 221))
+            paste_x = 0
+            paste_y = 0
+            if x < 0:
+                paste_x = -x
+                width = width + x
+                x = 0
+            if y < 0:
+                paste_y = -y
+                height = height + y
+                y = 0
+
+        # Render the region...
+        jpeg_data = image.renderJpegRegion(z, t, x, y, width, height,
+                                           level=level)
+        if jpeg_data is None:
+            return
+
+        i = StringIO(jpeg_data)
+        pil_img = Image.open(i)
+
+        # paste to canvas if needed
+        if canvas is not None:
+            canvas.paste(pil_img, (paste_x, paste_y))
+            pil_img = canvas
+
+        return pil_img
+
+    def get_panel_big_image(self, image, panel):
+        """Render the viewport region for BIG images"""
+
+        viewport_region = self.get_crop_region(panel)
+        rotation = int(panel.get('rotation', 0))
+        vp_x = viewport_region['x']
+        vp_y = viewport_region['y']
+        vp_w = viewport_region['width']
+        vp_h = viewport_region['height']
+        z = panel['theZ']
+        t = panel['theT']
+
+        # E.g. target is 300 dpi and width & height is '72 dpi'
+        # so we need image to be width * dpi/72 pixels
+        max_dpi = panel.get('max_export_dpi', 1000)
+        max_width = (panel['width'] * max_dpi) / 72
+
+        # Render a larger region than viewport, to allow for rotation...
+        if rotation != 0:
+            max_length = 1.5 * max(vp_w, vp_h)
+            extra_w = max_length - vp_w
+            extra_h = max_length - vp_h
+            viewport_region = {'x': vp_x - (extra_w/2),
+                               'y': vp_y - (extra_h/2),
+                               'width': vp_w + extra_w,
+                               'height': vp_h + extra_h}
+            max_width = max_width * (viewport_region['width'] / vp_w)
+
+        pil_img = self.render_big_image_region(image, z, t, viewport_region,
+                                               max_width)
+
+        # Optional rotation
+        if rotation != 0 and pil_img is not None:
+            w, h = pil_img.size
+            # How much smaller is the scaled image compared to viewport?
+            # This will be the same 'scale' used in render_big_image_region()
+            scale = float(w) / viewport_region['width']
```

Comment:
Could ``viewport_region['width']`` be ``0``?

Reply:
In theory it's possible to have a panel of zero width or height. In practice it's quite hard to create one (by dragging to resize) but I guess I should make it impossible since it does cause errors long before this point (export).